proposals/p2240.md

```diff
@@ -0,0 +1,1091 @@
+template <ConvertibleToString... Ts>
+std::string StrCat(const Ts&... params) {
+  std::string result;
+  result.reserve((params.Length() + ... + 0));
+  for... (auto param: params) {
+    result.append(param.ToString());
+  }
+  return result;
+}
+```
+
+</td>
+<td>
+
+With SE-0393 and SE-404
+
+```swift
+func StrCat<each T: ConvertibleToString>(_ param: repeat each T) -> String {
+  var len: Int64 = 0;
+  for param in repeat each param {
+    len += param.Length()
+  }
+  var result: String = ""
+  result.reserveCapacity(len)
+  for param in repeat each param {
+    result.append(param.ToString())
+  }
+  return result
+}
+```
+
+</td>
+</tr>
+</table>
+
+## Rationale
+
+Carbon needs variadics to effectively support
+[interoperation with and migration from C++](/docs/project/goals.md#interoperability-with-and-migration-from-existing-c-code),
+where variadic templates are fairly common. Variadics also make code
+[easier to read, understand, and write](/docs/project/goals.md#code-that-is-easy-to-read-understand-and-write),
+because some APIs (such as `printf`) can't be naturally expressed in terms of a
+fixed number of parameters.
+
+Furthermore, Carbon needs to support _generic_ variadics for the same reasons it
+needs to support generic non-variadic functions: for example,
+definition-checking makes APIs
+[easier to read, understand, and write](/docs/project/goals.md#code-that-is-easy-to-read-understand-and-write),
+and [easier to evolve](/docs/project/goals.md#software-and-language-evolution).
+Furthermore, the language as a whole is easier to understand and write code in
+if separate features like variadics and generics compose in natural ways, rather
+than being mutually exclusive.
+
+Variadics are also important for supporting
+[performance-critical software](/docs/project/goals.md#performance-critical-software),
+because variadic APIs can be more efficient than their non-variadic
+counterparts. For example, `StrCat` is fundamentally more efficient than
+something like a chain of `operator+` calls on `std::string`, because it does
+not need to materialize a series of partial results, and it can pre-allocate a
+buffer large enough for the final result.
+
+This proposal deviates somewhat from the principle that
+[all APIs are library APIs](/docs/project/principles/library_apis_only.md)
+because it treats tuples and packs as effectively built-in. As the document for
+this principle notes, tuples may be an exception to the principle because it is
+very hard to see how we could define them as library types without circularity.
+Packs have much the same problem, since a type representing a pack would itself
+have to be a variadic type. The need for both types to deeply support symbolic
+computation adds some major challenges here as well. However, if tuples could
+ever be library types, they would certainly need to be defined variadically, so
+in that sense this proposal may actually support that principle.
+
+## Alternatives considered
+
```

Comment:
Issue: `each(>=1)` syntax

This issue I'm more neutral on, but I think it is extra complexity we don't need if we don't have restrictive matching. It means that there is additional syntax to learn, more ways to write function signatures, and those ways are subtly different in ways that are hard to explain. It also has concerns around which `each` needs to marked with `(>=1)` and extra round trips if the user gets it wrong.